extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -0,0 +1,430 @@
+            this.response = response;
+        }
+
+        @Override
+        public void onStream(SdkPublisher<ByteBuffer> publisher) {
+            publisher.subscribe(new Request.Sub());
+        }
+
+        @Override
+        public void exceptionOccurred(Throwable error) {
+            producerFuture.completeExceptionally(error);
+        }
+
+        // --------------------------------------------------------------------------------------------------
+
+        private ByteBuffer get() throws ExecutionException, InterruptedException, TimeoutException {
+            if (released.isDone()) {
+                throw new IllegalStateException(String.format("Trying to get data after release, %s", requestStr()));
+            }
+            // Giving our own get() a bit of overhead - the clients should already be constructed with appropriate
+            // apiCallTimeout.
+            final long readNanos = instructions.readTimeout().plusMillis(100).toNanos();
+            return consumerFuture.get(readNanos, TimeUnit.NANOSECONDS);
+        }
+
+        private boolean isFragment(long fragmentIndex) {
+            return this.fragmentIndex == fragmentIndex;
+        }
+
+        private int requestLength() {
+            return (int) (to - from + 1);
+        }
+
+        private GetObjectRequest getObjectRequest() {
+            return GetObjectRequest.builder()
+                    .bucket(uri.bucket().orElseThrow())
+                    .key(uri.key().orElseThrow())
+                    .range("bytes=" + from + "-" + to)
+                    .build();
+        }
+
+        private String requestStr() {
+            if (uri != null) {
+                return String.format("ctx=%d ix=%d [%d, %d]/%d %s/%s", System.identityHashCode(S3ChannelContext.this),
+                        fragmentIndex, from, to, requestLength(), uri.bucket().orElseThrow(), uri.key().orElseThrow());
+            } else {
+                return String.format("ctx=%d ix=%d [%d, %d]/%d", System.identityHashCode(S3ChannelContext.this),
+                        fragmentIndex, from, to, requestLength());
+            }
+        }
+
+        // --------------------------------------------------------------------------------------------------
+
+        final class Sub implements Subscriber<ByteBuffer>, BiConsumer<Void, Throwable> {
+            private final CompletableFuture<ByteBuffer> localProducer;
+            private ByteBuffer buffer;
+            private Subscription subscription;
+
+            Sub() {
+                localProducer = producerFuture;
+                buffer = bufferPool.take(requestLength());
+                // 1. localProducer succeeds: whenComplete will executed when released
+                // 2. localProducer fails: whenComplete will executed asap
+                // 3. localProducer limbo (not expected): whenComplete will executed when released
+                CompletableFuture.allOf(localProducer, released).whenComplete(this);
```

Comment:
I'm worried about reachability here. You are sure we don't need to be.

Reply:
I originally said that some executor keeps it alive; that's not true. A CompleteableFuture knows it's dependents (and that's how the execution chain propagates when a root is completed):

```
jshell> import java.util.concurrent.CompletableFuture

jshell> var f1 = new CompletableFuture<>()
f1 ==> java.util.concurrent.CompletableFuture@b1bc7ed[Not completed]

jshell> var f2 = new CompletableFuture<>()
f2 ==> java.util.concurrent.CompletableFuture@30dae81[Not completed]

jshell> var f12 = CompletableFuture.allOf(f1, f2)
f12 ==> java.util.concurrent.CompletableFuture@4769b07b[Not completed]

jshell> f1
f1 ==> java.util.concurrent.CompletableFuture@b1bc7ed[Not completed, 1 dependents]

jshell> f2
f2 ==> java.util.concurrent.CompletableFuture@30dae81[Not completed, 1 dependents]
```

You see f1 and f2 know they have 1 dependent (in this case, the allOf future).

So in the code, as long as `released` is kept alive, our futures here will be kept alive.